### PR TITLE
Restore frames on nested break

### LINKED ISSUE #563: Existing from nested breakpoint doesn't resume to the correct frame
**Your environment**

* `ruby -v`: `3.1.0`
* `rdbg -v`: `dc93ef3ad7863a1bcfb34eb1bde78199fce9c504`

**Describe the bug**

#558 allows users to enter nested breakpoints. But after exiting the inner sub session, the debugger still stay at the inner frame.

**To Reproduce**

Script:

```rb
class Foo
  def self.bar(num)
    num
  end
end

binding.b
```

Command:

```
$ exe/rdbg -e 'b Foo.bar ;; c ;; bt ;; Foo.bar(10) ;; bt ;; c ;; bt' target.rb
```

Output:

```
❯ exe/rdbg -e 'b Foo.bar ;; c ;; bt ;; Foo.bar(10) ;; bt ;; c ;; bt' target.rb
[1, 7] in target.rb
=>   1| class Foo
     2|   def self.bar(num)
     3|     num
     4|   end
     5| end
     6|
     7| binding.b
=>#0    <main> at target.rb:1
(rdbg:commands) b Foo.bar
uninitialized constant Foo

Foo
^^^
#0  BP - Method (pending)  Foo.bar
(rdbg:commands) c
DEBUGGER:  BP - Method  Foo.bar at target.rb:2 is activated.
[2, 7] in target.rb
     2|   def self.bar(num)
     3|     num
     4|   end
     5| end
     6|
=>   7| binding.b
=>#0    <main> at target.rb:7
(rdbg:commands) bt
=>#0    <main> at target.rb:7
(rdbg:commands) Foo.bar(10)
[1, 7] in target.rb
     1| class Foo
     2|   def self.bar(num)
=>   3|     num
     4|   end
     5| end
     6|
     7| binding.b
=>#0    Foo.bar(num=10) at target.rb:3
  #1    <main> at (rdbg)/target.rb:1
  # and 3 frames (use `bt' command for all frames)

Stop by #0  BP - Method  Foo.bar at target.rb:2
(rdbg:commands) bt
=>#0    Foo.bar(num=10) at target.rb:3
  #1    <main> at (rdbg)/target.rb:1
  #2    [C] Binding#eval at ~/projects/debug/lib/debug/thread_client.rb:373
  #3    TracePoint.allow_reentry at <internal:trace_point>:151
  #4    <main> at target.rb:7
(rdbg:commands) c
10
(rdbg:commands) bt
=>#0    Foo.bar(num=10) at target.rb:3
  #1    <main> at (rdbg)/target.rb:1
  #2    [C] Binding#eval at ~/projects/debug/lib/debug/thread_client.rb:373
  #3    TracePoint.allow_reentry at <internal:trace_point>:151
  #4    <main> at target.rb:7
(rdbg) c    # continue command
```

**Expected behavior**

After exiting from the inner session (`target.rb:3`), the debugger should be back to the previous frame (`target.rb:7`).

**Additional context**

I think this can be solved by storing frames information (`@target_frames` and `@current_frame_index`) in stacks. I will open a PR for that shortly.

### PR DESCRIPTION
**Pull request by ko1**
fix #563

### CODE REVIEW

File: lib/debug/thread_client.rb [hunk 1]
```diff
@@ -359,6 +359,9 @@ def tp_allow_reentry
     end
 
     def frame_eval_core src, b
+      saved_target_frames = @target_frames
```
Comment (st0012) [inline lib/debug/thread_client.rb:362]: Doesn't this mean only 1 level of nesting is allowed?

Comment (ko1) [reply lib/debug/thread_client.rb:362]: please report if you have an suspicious case.

Comment (st0012) [reply lib/debug/thread_client.rb:362]: ok I can see it works even with 3 level of nesting

```
Stop by #2  BP - Method  Foo.baz at test.rb:11
(rdbg) bt    # backtrace command
=>#0    Foo.baz at test.rb:12
  #1    Foo.bar at (rdbg)/test.rb:1
  #2    [C] Binding#eval at ~/projects/debug/lib/debug/thread_client.rb:369
  #3    TracePoint.allow_reentry at <internal:trace_point>:151
  #4    Foo.bar at test.rb:8
  #5    Foo.foo at (rdbg)/test.rb:1
  #6    [C] Binding#eval at ~/projects/debug/lib/debug/thread_client.rb:369
  #7    TracePoint.allow_reentry at <internal:trace_point>:151
  #8    Foo.foo at test.rb:4
  #9    <main> at (rdbg)/test.rb:1
  #10   [C] Binding#eval at ~/projects/debug/lib/debug/thread_client.rb:369
  #11   TracePoint.allow_reentry at <internal:trace_point>:151
  #12   <main> at test.rb:17
(rdbg) c    # continue command
30
(rdbg) bt    # backtrace command
=>#0    Foo.bar at test.rb:8
  #1    Foo.foo at (rdbg)/test.rb:1
  #2    [C] Binding#eval at ~/projects/debug/lib/debug/thread_client.rb:369
  #3    TracePoint.allow_reentry at <internal:trace_point>:151
  #4    Foo.foo at test.rb:4
  #5    <main> at (rdbg)/test.rb:1
  #6    [C] Binding#eval at ~/projects/debug/lib/debug/thread_client.rb:369
  #7    TracePoint.allow_reentry at <internal:trace_point>:151
  #8    <main> at test.rb:17
(rdbg) c    # continue command
20
(rdbg) bt    # backtrace command
=>#0    Foo.foo at test.rb:4
  #1    <main> at (rdbg)/test.rb:1
  #2    [C] Binding#eval at ~/projects/debug/lib/debug/thread_client.rb:369
  #3    TracePoint.allow_reentry at <internal:trace_point>:151
  #4    <main> at test.rb:17
(rdbg) c    # continue command
10
(rdbg) bt    # backtrace command
=>#0    <main> at test.rb:17
(rdbg)
```

one question though: should these frames be displayed?

```
 #2    [C] Binding#eval at ~/projects/debug/lib/debug/thread_client.rb:369
 #3    TracePoint.allow_reentry at <internal:trace_point>:151
```

Comment (st0012) [reply lib/debug/thread_client.rb:362]: ah, it's to store the information in the current frame with local variables. I get it now.

Comment (ko1) [reply lib/debug/thread_client.rb:362]: > one question though: should these frames be displayed?

I don't think so. But eliminating them are not easy task and it is trivial (maybe it is rare to use nested breaks) so I left it.

Comment (st0012) [reply lib/debug/thread_client.rb:362]: I actually think it's easy to trigger nested breakpoints, but perhaps not intentionally. For example, having a `catch StandardError` first and make a typo in the REPL. I can even imagine people complaining about that 😂

**Pull request merged**